src/nnet3/nnet-chain-training.cc

```diff
@@ -68,23 +69,81 @@ void NnetChainTrainer::Train(const NnetChainExample &chain_eg) {
+    srand(srand_seed_ + num_minibatches_processed_);
+    ResetGenerators(nnet_);
+    TrainInternal(chain_eg, *computation, is_backstitch_step);
+    FreezeNaturalGradient(false, delta_nnet_); // un-freeze natural gradient
+    is_backstitch_step = false;
+    srand(srand_seed_ + num_minibatches_processed_);
+    ResetGenerators(nnet_);
+    TrainInternal(chain_eg, *computation, is_backstitch_step);
+  } else {
+    // conventional training
+    bool is_backstitch_step = false;
+    TrainInternal(chain_eg, *computation, is_backstitch_step);
+  }
+
+  num_minibatches_processed_++;
+}
+
+void NnetChainTrainer::TrainInternal(const NnetChainExample &eg,
+                                     const NnetComputation &computation,
+                                     bool is_backstitch_step) {
+  const NnetTrainerOptions &nnet_config = opts_.nnet_config;
+  NnetComputer computer(nnet_config.compute_config, computation,
                         *nnet_, delta_nnet_);
   // give the inputs to the computer object.
-  computer.AcceptInputs(*nnet_, chain_eg.inputs);
+  computer.AcceptInputs(*nnet_, eg.inputs);
   computer.Run();
 
-  this->ProcessOutputs(chain_eg, &computer);
+  this->ProcessOutputs(is_backstitch_step, eg, &computer);
   computer.Run();
 
-  UpdateParamsWithMaxChange();
+  // The configurations if doing conventional training with momentum
+  BaseFloat max_change_scale = 1.0, scale_adding = 1.0 - nnet_config.momentum,
+            scale_delta_nnet = nnet_config.momentum;
+  // The configurations if doing backstitch training
+  if (nnet_config.backstitch_training_scale > 0.0 && num_minibatches_processed_
+      % nnet_config.backstitch_training_interval == 0) {
+    if (is_backstitch_step) {
+      // max-change is scaled by backstitch_training_scale;
+      // delta_nnet is scaled by -backstitch_training_scale when added to nnet;
+      // delta_nnet itself is then zeroed
+      max_change_scale = nnet_config.backstitch_training_scale;
+      scale_adding = -nnet_config.backstitch_training_scale;
+      scale_delta_nnet = 0.0;
+    } else {
+      // max-change is scaled by 1 + backstitch_training_scale;
+      // delta_nnet is scaled by 1 + backstitch_training_scale when added to nnet;
+      // delta_nnet itself is then zeroed
+      max_change_scale = 1.0 + nnet_config.backstitch_training_scale;
+      scale_adding = 1.0 + nnet_config.backstitch_training_scale;
+      scale_delta_nnet = 0.0;
+    }
+  }
+  // Updates the parameters of nnet
+  bool success = UpdateNnetWithMaxChange(*delta_nnet_,
+      nnet_config.max_param_change, max_change_scale, scale_adding, nnet_,
+      &num_max_change_per_component_applied_, &num_max_change_global_applied_);
+  // Scales delta_nnet
+  if (success)
+    ScaleNnet(scale_delta_nnet, delta_nnet_);
+  else
+    ScaleNnet(0.0, delta_nnet_);
 }
 
-
-void NnetChainTrainer::ProcessOutputs(const NnetChainExample &eg,
+void NnetChainTrainer::ProcessOutputs(bool is_backstitch_step,
```

Comment:
I think it would  be better if you give this an argument is_backstitch_step2 and only added the
suffix _backstitch if that were true.  Because the regular "output" is the one we should really be plotting,
in most circumstances.  I'm a little unclear how you are obtaining your plots, because I assume you
need to pick up the one with the _backstitch suffix, given the way the code currently works.

Note: once you split TrainInternal() into two functions, doing this will be easier, because in the calling
code you'll know whether you are doing backstitch or not.